source/common/modules/markdown-editor/plugins/highlight-whitespace.ts

```diff
@@ -16,26 +16,82 @@ import {
   type Extension,
   Compartment,
   StateEffect,
-  EditorState
+  EditorState,
+  RangeSetBuilder
 } from '@codemirror/state'
 import {
   highlightWhitespace as hw,
-  highlightTrailingWhitespace as htw
+  highlightTrailingWhitespace as htw,
+  Decoration,
+  ViewPlugin,
+  type ViewUpdate,
+  type DecorationSet,
+  type EditorView,
+  WidgetType
 } from '@codemirror/view'
 import { configUpdateEffect } from '../util/configuration'
 
 const extensionCompartment = new Compartment()
 
 /**
- * A dark mode configuration effect. Pass this to the editor whenever you want
- * to exchange the light or dark themes that the editor uses, or to switch
- * between the light and dark mode.
+ * A widget which adds a pilcrow, '¶', to the end of lines.
+ */
+class PilcrowWidget extends WidgetType {
+  toDOM () {
+    const span = document.createElement('span')
+    span.className = 'cm-pilcrow'
+    span.textContent = '¶'
+    span.style = 'color: #666'
```

Comment:
Mh, while I think the pilcrow symbol shouldn't be CSS as we discussed, I do believe that the coloring should be moved to a theme, just so that it's a bit more colocated (see the Typewriter plugin). It's gonna be a bit empty, but I think it makes the code more streamlined and easier to understand!